### PR TITLE
Add support for sysctl-based evdev property enumeration

### PR DESCRIPTION
**Pull request by valpackett**
A big security advantage of Wayland is not allowing applications to read input devices all the time. Having `/dev/input/*` accessible to the user account subverts this advantage.

libudev-devd was opening the evdev devices to detect their types (mouse, keyboard, touchpad, etc). This does not work when `/dev/input/*` is inaccessible.

With the kernel exposing this information as sysctls, we can work without `/dev/input/*` access, preserving the Wayland security model.

Even more: thanks to libcasper, it was pretty easy to add Capsicum support, so now my capsicumized Weston can hotplug devices! :) (UPD: left that out for now)

- [D18694](https://reviews.freebsd.org/D18694) base patch that adds the sysctls

### CODE REVIEW

**Comment by zeising** [comment]
This needs to be made (if it isn't already) so that this works on FreeBSD versions both with and without the sysctls in your base patch.

It looks like this is already done.

**Comment by zeising** [comment]
I have no objections, but would like to see where the FreeBSD base part of this goes first.

**Comment by zeising** [comment]
Hm. I probably made a mistake there, I only intended to add myself to reviewers, not remove the other ones.

**Comment by valpackett** [comment]
Yeah, it is compatible with the old way, of course:

```c
goto use_ioctl;
```

Just retested with an old kernel, no problems, the old one requires `chmod g+rw /dev/input/*` while the new one doesn't.

**Comment by zeising** [comment]
> Yeah, it is compatible with the old way, of course:
> 
> ```c
> goto use_ioctl;
> ```
> 
> Just retested with an old kernel, no problems, the old one requires `chmod g+rw /dev/input/*` while the new one doesn't.

Yeah, I noticed that when I took a closer look at the code.

**Comment by zeising** [comment]
Is this current with the changes that just went into the kernel, or does it need adjusting?

**Comment by valpackett** [comment]
"input_id" was renamed to just "id". I pushed the corresponding change here. Will re-test everything later today when I get home.

**Comment by valpackett** [comment]
Yep, everything works great, this is good to go.

File: udev-utils.c [hunk 1]
```diff
@@ -322,6 +323,50 @@ create_evdev_handler(struct udev_device *ud)
...
+	if (sysctlbyname(mib, phys, &len, NULL, 0) < 0)
+		goto use_ioctl;
+
+	snprintf(mib, sizeof(mib), "kern.evdev.input.%s.id", unit);
+	len = sizeof(id);
+	if (sysctlbyname(mib, &id, &len, NULL, 0) < 0)
+		goto use_ioctl;
+
+	snprintf(mib, sizeof(mib), "kern.evdev.input.%s.key_bits", unit);
+	len = sizeof(key_bits) * sizeof(key_bits[0]);
```
Comment (wulf7) [inline udev-utils.c]: It should be just a size_of(key_bits)?

Comment (valpackett) [reply udev-utils.c]: oops, nice catch

**Comment by valpackett** [comment]
ping

**Comment by valpackett** [comment]
rebased. wooo CI works nice

**Comment by valpackett** [comment]
@zeising ping

**Comment by valpackett** [comment]
@johalun @zeising ping

**Comment by johalun** [comment]
pong! Looks like things are in the kernel. Are we good to merge this and update the port?

**Comment by valpackett** [comment]
Yeah the things have been in kernel (including -STABLE) for a while.

**Comment by zeising** [comment]
Sorry, I've been focusing on other things.

**Pull request merged**